### PR TITLE
Add support for different journal modes

### LINKED ISSUE #2040: random database lock when sqlite database
This is a summary issue since there are so many issues to report the database lock when using sqlite. I will close all other issues and please continue in this issue.

### PR DESCRIPTION
**Pull request by jonasfranz**
Fixes #2040 (partially)

Add the ability to set the journal mode used by SQLite3. The new default journal mode is WAL, since it produces more less database locked errors (nearly no errors).

Learn more about WAL: https://www.sqlite.org/wal.html 

https://github.com/mattn/go-sqlite3/issues/569#issuecomment-393202136

Maybe should we also add `SetMaxOpenConnection(1)`. But I'm not sure about that.

### CODE REVIEW

File: modules/indexer/issue.go [hunk 1]
```diff
@@ -64,7 +64,7 @@ func InitIssueIndexer(populateIndexer func() error) {
 		log.Fatal(4, "InitIssuesIndexer: create index, %v", err)
 	}
 	if err = populateIndexer(); err != nil {
-		log.Fatal(4, "InitIssueIndexer: populate index, %v", err)
+		//log.Fatal(4, "InitIssueIndexer: populate index, %v", err)
```
Comment (lunny) [inline modules/indexer/issue.go:67]: why comment this line?

**Comment by lunny** [comment]
@JonasFranzDEV Since that, could you enable `make test-sqlite`?

File: custom/conf/app.ini.sample [hunk 2]
```diff
@@ -213,6 +213,8 @@ SSL_MODE = disable
 PATH = data/gitea.db
 ; For "sqlite3" only. Query timeout
 SQLITE_TIMEOUT = 500
+; For "sqlite3" only. Journal mode
+JOURNAL_MODE = WLA
```
Comment (Bwko) [inline custom/conf/app.ini.sample]: WLA -> WAL (https://www.sqlite.org/pragma.html#pragma_journal_mode)

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 3]
```diff
@@ -129,6 +129,9 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 - `PASSWD`: **\<empty\>**: Database user password. Use \`your password\` for quoting if you use special characters in the password.
 - `SSL_MODE`: **disable**: For PostgreSQL only.
 - `PATH`: **data/gitea.db**: For SQLite3 only, the database file path.
+- `JOURNAL_MODE` **WLA**: For SQLite3 only, the journal mode. Use `DELETE` mode if you're running Gitea on a network file system
```
Comment (Bwko) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]: Same

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 4]
```diff
@@ -129,6 +129,9 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
...
+or your operating system does not support VFS. Please checkout the [WAL documentation](https://www.sqlite.org/wal.html) to learn more. WLA
```
Comment (Bwko) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]: Same

**Comment by lafriks** [comment]
Info about journal mode should be also added in admin settings display page

**Comment by lunny** [comment]
The wal mode maybe an enforce mode not an option.

**Comment by ghost** [comment]
Some files are missing in the last commit (c58e2c5) e.g. `sqlite3_func_crypt.go`

In addition, I think the database connection need be closed before Gitea exits. Otherwise, `gitea.db-shm` and `gitea.db-wal` files never gets deleted.

https://www.sqlite.org/wal.html#avoiding_excessively_large_wal_files
> When the last connection to a database closes, that connection does one last checkpoint and then deletes the WAL and its associated shared-memory file, to clean up the disk.

https://www.sqlite.org/tempfiles.html#write_ahead_log_wal_files
> However, if the last connection does not shutdown cleanly, the WAL file will remain in the filesystem and will be automatically cleaned up the next time the database is opened.

But I don't know why it hasn't been deleted or reset to 0KB.

**Comment by jonasfranz** [comment]
I've tried it locally and it does not improve the tests at any point. I think I should close this PR since it does not offer benefits to the user. Also setting max open connections to 1 results in the test hanging up.

**Pull request closed without merging**